cub/cub/thread/thread_scan.cuh

```diff
@@ -164,6 +164,87 @@ ThreadScanExclusive(T (&input)[LENGTH], T (&output)[LENGTH], ScanOp scan_op, T p
   return ThreadScanExclusive<LENGTH>((T*) input, (T*) output, scan_op, prefix, apply_prefix);
 }
 
+/**
+ * @brief Perform a sequential exclusive prefix scan over the first @p valid_items elements
+ *        of the statically-sized @p input array, seeded with the specified @p prefix.
+ *        The aggregate is returned.
+ *
+ * @tparam Input
+ *   <b>[inferred]</b> The data type holding the input data having member
+ *   <tt>operator[](int i)</tt> and must be statically-sized (<tt>size()</tt> method or static array)
+ *
+ * @tparam Output
+ *   <b>[inferred]</b> The data type to hold the scan result having member
+ *   <tt>operator[](int i)</tt> and must be statically-sized (<tt>size()</tt> method or static array)
+ *
+ * @tparam ScanOp
+ *   <b>[inferred]</b> Binary scan operator type having member
+ *   <tt>T operator()(const T &a, const T &b)</tt>
+ *
+ * @tparam PrefixT
+ *   <b>[inferred]</b> The data type of the prefix.
+ *
+ * @param[in] input
+ *   Array=like input
+ *
+ * @param[out] output
+ *   Array=like output (may be aliased to @p input)
+ *
+ * @param[in] scan_op
+ *   Binary scan operator
+ *
+ * @param[in] valid_items
+ *   Number of valid items
+ *
+ * @param[in] prefix
+ *   Prefix to seed scan with
+ *
+ * @param[in] apply_prefix
+ *   Whether or not the calling thread should apply its prefix.
+ *   (Handy for preventing thread-0 from applying a prefix.)
+ */
+template <typename Input,
+          typename Output,
+          typename ScanOp,
+          typename PrefixT,
+          typename ValueT = _CUDA_VSTD::iter_value_t<Input>,
+          typename AccumT = _CUDA_VSTD::__accumulator_t<ScanOp, ValueT, PrefixT>>
+_CCCL_DEVICE _CCCL_FORCEINLINE void ThreadScanExclusivePartial(
+  Input& input, Output& output, ScanOp scan_op, int valid_items, PrefixT prefix, bool apply_prefix = true)
+{
+  using namespace cub::detail;
+  static_assert(is_fixed_size_random_access_range_v<Input>,
+                "Input must support the subscript operator[] and have a compile-time size");
+  static_assert(is_fixed_size_random_access_range_v<Output>,
+                "Output must support the subscript operator[] and have a compile-time size");
+  static_assert(has_binary_call_operator<ScanOp, ValueT>::value,
+                "ScanOp must have the binary call operator: operator(ValueT, ValueT)");
+  constexpr auto length = static_size_v<Input>;
+  static_assert(static_size_v<Output> == length);
+  AccumT inclusive = input[0];
+  if (valid_items > 0)
+  {
+    if (apply_prefix)
+    {
+      inclusive = scan_op(prefix, inclusive);
+    }
+    output[0] = prefix;
+  }
+
+  AccumT exclusive = inclusive;
+
+  _CCCL_PRAGMA_UNROLL_FULL()
```

Comment:
Q: Have you measures that unrolling has a positive performance impact? We observed regressions with full unrolling in the near past and it also increases compile-time.

Reply:
I imagine this will depend on the context this will appear in. Again, I would advise to not write dedicated benchmarks for now but instead wait for this utility to be integrated in the `DeviceScan` and benchmark the influence of unrolling in that context.

Reply:
Sorry, realized just now that I hadn't flushed my previous comment but have done so now 🙂